File: 23-Merge k Sorted Lists.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

import heapq
class Solution(object):
    def mergeKLists(self, lists):
        """
        :type lists: List[ListNode]
        :rtype: ListNode
        """
# Sort all lists O(NlogN)
        res = []
        for node in lists:
            while node:
                res.append(node.val)
                node = node.next
        res.sort()
        curr = dummy = ListNode(0)
        for num in res:
            dummy.next = ListNode(num)
            dummy = dummy.next
        return curr.next
    
# priority queue O(Nlogk)
        
        curr = dummy = ListNode(0)
        minheap = []
        for node in lists:
            if node:
                heapq.heappush(minheap,(node.val, node))
        
        while minheap:
            value, node = heapq.heappop(minheap)
            curr.next = node
            curr = curr.next
            if curr.next:
                heapq.heappush(minheap, (curr.next.val, curr.next))
                
        return dummy.next

# Merge with Divide And Conquer O(Nlogk)
        
        def merge2list(l1, l2):
            dummy = curr = ListNode(0)
            while l1 and l2:
                if l1.val > l2.val:
                    curr.next = l2
                    l2 = l2.next
                else:
                    curr.next = l1
                    l1 = l1.next  
                curr = curr.next
            if l2:
                curr.next = l2
            else:
                curr.next = l1
            dummy2 = dummy
            while dummy2:
                dummy2 = dummy2.next
            return dummy.next
        
        def mergeAll(lists, l, r):
            if l>r:
                return None
            elif l == r:
                return lists[l]
            elif l + 1 == r:
                return merge2list(lists[l], lists[r])
            else:
                mid = l + (r-l)//2
                left = mergeAll(lists, l, mid)
                right = mergeAll(lists, mid+1, r)
                return merge2list(left, right)

        return mergeAll(lists, 0, len(lists)-1)

File: JavaScript/23-Merge k Sorted Lists.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode[]} lists
 * @return {ListNode}
 */
var mergeKLists = function(lists) {
    if (lists.length < 1) return null;
    
    let mergeTwo = (one, two) =>{
        let dummy = new ListNode(0);
        let curr = dummy;
        while (one && two){
            if (one.val<=two.val){
                curr.next = new ListNode(one.val);
                one = one.next;
            } else{
                curr.next = new ListNode(two.val);
                two = two.next;
            }
            curr = curr.next;
        }
        curr.next = one || two;
        return dummy.next;
    }
    
    let merge = (lists) =>{
        if (lists.length <= 1) return lists;
        let arr = [];
        for (i=0; i<lists.length; i+=2){
            arr.push(mergeTwo(lists[i], lists[i+1]));
        }
        if (i < lists.length-1) arr.push(lists[lists.length-1])
        return merge(arr)
    }
    
    return merge(lists)[0] ? merge(lists)[0]:null
    
};